### ISSUE TITLE
NSRangeException from NSTaggedPointerString on iOS 9

### ISSUE THREAD

**Issue by HimaniM** [card-io/card.io-iOS-SDK#158]
I checked by implementing all versions of cardio like version 5.2.2 and 5.1.0 But issues is still persist.
In ios 9,
    CardIOPaymentViewController *scanViewController = [[CardIOPaymentViewController alloc] initWithPaymentDelegate:self];
    scanViewController.modalPresentationStyle = UIModalPresentationFormSheet;
    [self presentViewController:scanViewController animated:YES completion:nil];

After above lines the app crashes in ios 9 not in ios 8.

Logs
*** Terminating app due to uncaught exception 'NSRangeException', reason: '-[NSTaggedPointerString getCharacters:range:]: Range {0, 7} out of bounds; string length 6'
*** First throw call stack:
(0x1867a4f5c 0x19acfbf80 0x1867a4ea4 0x18678c4e4 0x1866970d4 0x18678cc90 0x1004e5f0c 0x1866dd124 0x1004e5c58 0x1004e67ac 0x1004e6958 0x1004c0a54 0x18bcfb610 0x18bcfb194 0x18c08a830 0x18c049734 0x18c046294 0x18bfa4828 0x18bfb0dc8 0x18bced1c8 0x18675bc30 0x1867599d4 0x186759e04 0x186688dc0 0x191600088 0x18bd62f44 0x10007d3cc 0x19b51a8b8)
libc++abi.dylib: terminating with uncaught exception of type NSException

**Comment by dgoldman-ebay** [comment]
@HimaniM could you also list the steps you followed as a user, leading to the crash? E.g., did you successfully scan a card? Or switch to the Manual Entry screen? Make a manual entry?

Also, simulator or device? Which device (actual or simulated)?

**Comment by HimaniM** [comment]
yes sure, I am using ios 9.0 version on device. 
I have a button. when i tap it crashes. No screen is visible to scan and all before that only it crashes.

**Comment by dgoldman-ebay** [comment]
And which device? (E.g., iPhone 5s, iPhone 6+, iPad Retina, etc.)

**Comment by dgoldman-ebay** [comment]
Oh, also -- could you please try building and running the Sample App that comes with the card.io SDK to determine whether the same crash occurs there?

**Comment by HimaniM** [comment]
The device i tried are iPod 5 and iphone 6 both 64 bit. The strange thing that sample app is working fine in both devices.

**Comment by HimaniM** [comment]
When i copied same Card io folder (library and .h files) from sample app and run my application its not working for ios 9 only.

**Comment by dgoldman-ebay** [comment]
@HimaniM It is interesting that the Sample App runs successfully but your app does not.

Unfortunately, I do not currently have a 64-bit iOS 9 device on hand myself, so right at this moment I don't really have a way to pursue this.

I would suggest comparing the relevant parts of your code, line by line, with the code of the Sample App. Hopefully you will be able to find the difference that leads to this crash.

**Comment by HimaniM** [comment]
Atlast i found the solution, thats why thought to share with all

From ios 9, Need to preload it first before opening any CardIO related UI.

Just write 
    [CardIOUtilities preload];
in viewWillAppear of viewController from where you are going to present CardIOPaymentViewController.

Thanks !!

**Comment by dgoldman-ebay** [comment]
@HimaniM thanks for reporting your solution!

(I'm surprised that not calling `[CardIOUtilities preload]` leads to this crash. I'll look into that when I have a moment.)

**Comment by mabril** [comment]
Same problem here with iOS 9. Preload does not fix it...

We are deactivating it as it's stopping our sales

**Comment by josharian** [comment]
Details will help. Do you have symbolized backtraces? Do you know what localization is in use, and what the user was doing? Reproduction instructions?

**Comment by NKorchagin** [comment]
Hi. So, is this bug still not fixed since November? App crashes with NSRangeException. CardIO version - 5.2.2
Reproducible: iPhone6/iOS9.2, iPhone6+/iOS9.1, iPhone6s/iOS9.2
NOT Reproducible: iPod5/ iOS9.1, iPhone5/iOS8.4, iPhone4S/iOS9, iPhone5/iOS9.2

<Error>: *** Terminating app due to uncaught exception 'NSRangeException', reason: '-[NSTaggedPointerString getCharacters:range:]: Range
{0, 7}
out of bounds; string length 6'

**Comment by bluk** [comment]
@NKorchagin Can you provide some additional details like what localization is in use, what the user is doing and detailed reproduction steps? Example code to reproduce your crash may also be useful.

**Comment by NKorchagin** [comment]
1) We used different device localizations:
US, RU, etc.
Same issue for any of them. Depending on device model, not its localization.

2) We have scan button. User just press it, nothing happened, nothing presented, immediate crash with NSRangeException in log.

3) Scan button triggers selector with this code:
```
[someUITextView becomeFirstResponder];
CardIOPaymentViewController *scanViewController = [[CardIOPaymentViewController alloc] initWithPaymentDelegate:self];
scanViewController.collectCVV = NO;
scanViewController.collectExpiry = NO;
scanViewController.collectPostalCode = NO;
scanViewController.disableManualEntryButtons = YES;
[somePresentationController presentViewController:scanViewController animated:YES completion:nil];
```

4) That's it. We have no more CardIO-related code in app except 2 delegate callbacks:
-userDidCancelPaymentViewController
-userDidProvideCreditCardInfo

**Comment by NKorchagin** [comment]
After updating CardIO version to 5.3 we made regression test on this issue with same devices and version. Seems that this bug is fixed in 5.3. We did not have any crashes.

**Comment by bluk** [comment]
@NKorchagin Thanks for following up! Still had trouble reproducing your issue on a couple of the hardware/versions you specified, but I'm glad that 5.3 resolves the issue for you. (There were a few changes to some translation strings but I'm not aware of any other changes that would directly impact this issue.)
